src/view/src/main/java/airqo/services/BigQueryApiImpl.java

```diff
@@ -113,6 +114,20 @@ public List<GraphInsight> getInsights(Date startDateTime, Date endDateTime, Stri
 		return insights;
 	}
 
+	@Override
+	@Cacheable(value = "bigQueryInsightsCache", cacheNames = {"bigQueryInsightsCache"}, unless = "#result.isEmpty()")
+	public List<GraphInsight> getInsights(Date startDateTime, Date endDateTime, String siteId) {
+		log.info(String.format("%s is not cached ", siteId));
+		return queryInsights(startDateTime, endDateTime, siteId);
+	}
+
+	@Override
+	@CachePut(value = "bigQueryInsightsCache", cacheNames = {"bigQueryInsightsCache"}, unless = "#result.isEmpty()")
+	public List<GraphInsight> cacheInsights(Date startDateTime, Date endDateTime, String siteId) {
+		log.info(String.format("Updating cache for site => %s, Start date time =>  %s, End date time =>  %s,", siteId, startDateTime, endDateTime));
+		return queryInsights(startDateTime, endDateTime, siteId);
+	}
```

Comment:
Updating cache

Reply:
Thanks for this @NoahNsimbe , this is a much welcome optimisation.
I really like the separation of concerns.

- Is the update of this cache triggered by user actions?
- Have we catered for all possible scenarios associated with our cache identifiers: siteId, endTime and startTime? Does any of them missing still generate desired result? Does the ENVIRONMENT matter in this case in case the caching system is all the same?
- 
Thanks again.

Reply:
**Is the update of this cache triggered by user actions?**

`Triggered by a notification from the message broker.`

**Have we catered for all possible scenarios associated with our cache identifiers: siteId, endTime and startTime? Does any of them missing still generate desired result?** 

`Yes, we have catered for the identifies. These values don't change for a period of one month. The mobile app is designed to always display data for a month.`

Does the ENVIRONMENT matter in this case in case the caching system is all the same?

`This is handled by kubernetes. We have separate clusters for the different environments.`

@Baalmart